### PR TITLE
[refactor] Controllerserver (part 1) - Move existing helpers and add better comments

### PR DESCRIPTION
**Pull request by komer3**
### General:

* [ ] Have you removed all sensitive information, including but not limited to access keys and passwords?
* [ ] Have you checked to ensure there aren't other open or closed [Pull Requests](../../pulls) for the same bug/feature/question?

### Pull Request Guidelines:

1. [ ] Does your submission pass tests?
1. [ ] Have you added tests? 
1. [ ] Are you addressing a single feature in this PR? 
1. [ ] Are your commits atomic, addressing one change per commit?
1. [ ] Are you following the conventions of the language? 
1. [ ] Have you saved your large formatting changes for a different PR, so we can focus on your work?
1. [ ] Have you explained your rationale for why this feature is needed? 
1. [ ] Have you linked your PR to an [open issue](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/)

### CODE REVIEW

File: internal/driver/controllerserver_helper.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	volumes, err := s.client.ListInstanceVolumes(ctx, instance.ID, nil)
+	if err != nil {
+		return false, errInternal("list instance volumes: %v", err)
+	}
+
+	// Return true if the number of attached volumes is less than the limit
+	return len(volumes) < limit, nil
+}
+
+// maxVolumeAttachments calculates the maximum number of volumes that can be attached to a Linode instance,
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:104]: while we are at it, can you remove the limits and limits_tests modules? I believe this is being used in favor of the function in limits

Comment (amold1) [reply internal/driver/controllerserver_helper.go:104]: Also can we rename this to `maxAllowedVolumeAttachments`?

File: internal/driver/controllerserver_helper.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+	}
+
+	// Convert the reported memory from MB to bytes
+	memBytes := uint(instance.Specs.Memory) << 20
+	return maxVolumeAttachments(memBytes) - len(disks), nil
+}
+
+// attemptGetContentSourceVolume retrieves information about the Linode volume to clone from.
+// It returns a LinodeVolumeKey if a valid source volume is found, or an error if the source is invalid.
+func (cs *ControllerServer) attemptGetContentSourceVolume(ctx context.Context, contentSource *csi.VolumeContentSource) (*linodevolumes.LinodeVolumeKey, error) {
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:128]: can we rename these methods and remove the `attempt` prefix?

File: internal/driver/controllerserver_helper.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+		return nil, errUnsupportedVolumeContentSource
+	}
+
+	sourceVolume := contentSource.GetVolume()
+	if sourceVolume == nil {
+		return nil, errNoSourceVolume // Return error if no source volume is specified
+	}
+
+	// Parse the volume ID from the content source
+	volumeInfo, err := linodevolumes.ParseLinodeVolumeKey(sourceVolume.GetVolumeId())
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:147]: volumeInfo being a pointer, we should add a nilcheck for it

File: internal/driver/controllerserver_helper.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+	if err != nil {
+		return nil, errInternal("parse volume info from content source: %v", err)
+	}
+
+	// Retrieve the volume data using the parsed volume ID
+	volumeData, err := cs.client.GetVolume(ctx, volumeInfo.VolumeID)
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:153]: same here

File: internal/driver/controllerserver_helper.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+	}
+
+	volumes, err := cs.client.ListVolumes(ctx, linodego.NewListOptions(0, string(jsonFilter)))
+	if err != nil {
+		return nil, errInternal("list volumes: %v", err)
+	}
+
+	// Raise an error if more than one volume with the same label exists
+	if len(volumes) > 1 {
+		return nil, errAlreadyExists("volume %q already exists", label)
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:186]: this error message doesnt match the comment above. if multiple volumes are found, the error message should reflect that

**Review by amold1 — COMMENTED** [review]
great job on the clearer comments in the code! 

LGTM... just a few naming and nilcheck comments

File: internal/driver/controllerserver_helper.go (hunk 5 above)
Comment (komer3) [reply internal/driver/controllerserver_helper.go:186]: ah good catch!

File: internal/driver/controllerserver_helper.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+	// Parse the volume ID from the content source
+	volumeInfo, err := linodevolumes.ParseLinodeVolumeKey(sourceVolume.GetVolumeId())
+	if err != nil || volumeInfo == nil {
+		return nil, errInternal("parse volume info from content source: %v", err)
+	}
+
+	// Retrieve the volume data using the parsed volume ID
+	volumeData, err := cs.client.GetVolume(ctx, volumeInfo.VolumeID)
+	if err != nil || volumeData == nil {
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:154]: I think we should have separate checks. If err is nil and volumeData is nil as well, the error message will not be of much help
```
get volume <nil>: <nil>"
```

File: internal/driver/controllerserver_helper.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+	}
+
+	sourceVolume := contentSource.GetVolume()
+	if sourceVolume == nil {
+		return nil, errNoSourceVolume // Return error if no source volume is specified
+	}
+
+	// Parse the volume ID from the content source
+	volumeInfo, err := linodevolumes.ParseLinodeVolumeKey(sourceVolume.GetVolumeId())
+	if err != nil || volumeInfo == nil {
```
Comment (amold1) [inline internal/driver/controllerserver_helper.go:148]: same here

**Review by amold1 — APPROVED** [review]

**Review by eljohnson92 — APPROVED** [review]

**Pull request merged**